csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +261,386 @@ int64_t getMaxPersistentBatch(
+    const int64_t total_reduction_numel,
+    const int64_t total_iteration_numel,
+    const int64_t inner_most_dimension_numel,
+    const int64_t n_tensor_inputs,
+    const int64_t max_input_dtype_size,
+    const int64_t max_persistent_buffer_size,
+    const size_t vectorize_factor,
+    const bool project_to_input,
+    const PrimDataType index_type,
+    const bool has_exp_op,
+    const bool has_rng_op) {
+  // Define two free parameters used in this heuristic.
+  // register_overhead is all registers except those for the persistent
+  // buffers. The register in each thread = register_overhead +
+  // persistent_buffer_size / bytes_per_register
+  // Current values are based on tests of sofmax, layer_norm, softmax_dropout,
+  // dropout_layer_norm on A100 & H100. It directly affects maxregcount passed
+  // to NVRTC and influences the occupancy.
+  const int64_t register_overhead = has_exp_op ? 32l : 16l;
+
+  // Target occupancy required to hide memory latency.
+  // Used to calculate the maximum register count each thread can use.
+  // Used to calculate the maximum persistent batch size.
+  // Current value is based on tests of sofmax, layer_norm, softmax_dropout,
+  // dropout_layer_norm on A100 & H100.
+  const int64_t target_warps_per_sm = 28l;
+
+  // device properties
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t threads_per_warp = (int64_t)dev_prop->warpSize;
+  const int64_t max_threads_in_block = (int64_t)dev_prop->maxThreadsPerBlock;
+  const int64_t max_threads_per_sm =
+      (int64_t)dev_prop->maxThreadsPerMultiProcessor;
+  const int64_t device_multiprocessor_count =
+      (int64_t)dev_prop->multiProcessorCount;
+
+  // alwasy use [vectorize_factor]
+  NVF_CHECK(
+      inner_most_dimension_numel % vectorize_factor == 0,
+      "Element count should be divisible by vectorize factor! count= ",
+      inner_most_dimension_numel,
+      ", vectorize_factor= ",
+      vectorize_factor);
+  const int64_t parallel_after_vectorize =
+      inner_most_dimension_numel / (int64_t)vectorize_factor;
+
+  // try to use at least 4 warps per block
+  const int64_t min_threads_per_block = 4l * threads_per_warp;
+
+  // set the min persistent buffer size to avoid requesting
+  // a block size larger than device limit
+  const int64_t batches_per_block_inner_reduction_min =
+      ceilDiv(parallel_after_vectorize, max_threads_in_block);
+
+  // set the max persistent batch size to avoid low occupancy
+  // don't increase persistent batch size until we have enough threads
+  const int64_t pbs_max_1 =
+      ceilDiv(parallel_after_vectorize, min_threads_per_block);
+  // derived the maximum persistent batch size from the target occupancy
+  const int64_t buffer_bytes_per_batch = max_persistent_buffer_size /
+      total_reduction_numel * (int64_t)vectorize_factor;
+  const int64_t target_threads_per_sm =
+      std::min(target_warps_per_sm * threads_per_warp, max_threads_per_sm);
```

Comment:
sorry I think this is what I meant earlier.

i.e. if we have `target_warps_per_sm * threads_per_warp > max_threads_per_sm`, it should be an assert because we are setting an unrealistic `target_warps_per_sm`...